File: src/main/java/com/braydenoneal/block/entity/ModBlockEntities.java
package com.braydenoneal.block.entity;

import com.braydenoneal.Blogic;
import com.braydenoneal.block.ModBlocks;
import net.fabricmc.fabric.api.object.builder.v1.block.entity.FabricBlockEntityTypeBuilder;
import net.minecraft.block.Block;
import net.minecraft.block.entity.BlockEntity;
import net.minecraft.block.entity.BlockEntityType;
import net.minecraft.registry.Registries;
import net.minecraft.registry.Registry;
import net.minecraft.util.Identifier;

public class ModBlockEntities {
    public static final BlockEntityType<RedstoneReaderBlockEntity> REDSTONE_READER_BLOCK_ENTITY =
            register("redstone_reader", RedstoneReaderBlockEntity::new, ModBlocks.REDSTONE_READER);
    public static final BlockEntityType<RedstoneWriterBlockEntity> REDSTONE_WRITER_BLOCK_ENTITY =
            register("redstone_reader", RedstoneWriterBlockEntity::new, ModBlocks.REDSTONE_WRITER);

    private static <T extends BlockEntity> BlockEntityType<T> register(
            String name,
            FabricBlockEntityTypeBuilder.Factory<? extends T> entityFactory,
            Block... blocks
    ) {
        Identifier id = Identifier.of(Blogic.MOD_ID, name);
        return Registry.register(Registries.BLOCK_ENTITY_TYPE, id, FabricBlockEntityTypeBuilder.<T>create(entityFactory, blocks).build());
    }

    public static void initialize() {
    }
}

File: src/main/kotlin/com/braydenoneal/block/entity/ModBlockEntities.kt
package com.braydenoneal.block.entity

import com.braydenoneal.Blogic
import com.braydenoneal.block.ModBlocks
import net.fabricmc.fabric.api.`object`.builder.v1.block.entity.FabricBlockEntityTypeBuilder
import net.fabricmc.fabric.api.screenhandler.v1.ExtendedScreenHandlerType
import net.minecraft.block.Block
import net.minecraft.block.BlockState
import net.minecraft.block.entity.BlockEntity
import net.minecraft.block.entity.BlockEntityType
import net.minecraft.entity.player.PlayerInventory
import net.minecraft.registry.Registries
import net.minecraft.registry.Registry
import net.minecraft.screen.ScreenHandlerType
import net.minecraft.util.Identifier
import net.minecraft.util.math.BlockPos

class ModBlockEntities {
    companion object {
        val CONTROLLER_BLOCK_ENTITY = register("controller", FabricBlockEntityTypeBuilder.Factory { pos: BlockPos, state: BlockState -> ControllerBlockEntity(pos, state) }, ModBlocks.CONTROLLER)

        val CONTROLLER_SCREEN_HANDLER: ExtendedScreenHandlerType<ControllerScreenHandler, BlockPos> =
            Registry.register<ScreenHandlerType<*>, ExtendedScreenHandlerType<ControllerScreenHandler, BlockPos>>(
                Registries.SCREEN_HANDLER,
                Identifier.of("blogic", "controller_block"),
                ExtendedScreenHandlerType<ControllerScreenHandler, BlockPos>({ syncId: Int, playerInventory: PlayerInventory, pos: BlockPos ->
                    ControllerScreenHandler(syncId, playerInventory, pos)
                }, BlockPos.PACKET_CODEC)
            )

        private fun <T : BlockEntity> register(
            name: String,
            entityFactory: FabricBlockEntityTypeBuilder.Factory<out T>,
            vararg blocks: Block
        ): BlockEntityType<T> {
            val id = Identifier.of(Blogic.MOD_ID, name)
            return Registry.register(
                Registries.BLOCK_ENTITY_TYPE, id,
                FabricBlockEntityTypeBuilder.create<T>(entityFactory, *blocks).build()
            )
        }

        fun initialize() {
        }
    }
}
